### PR TITLE
Implement the determinations of RFD 285

### PR DESCRIPTION
**Pull request by ahl**
This implements the renames described in 285 to the best of my ability. I would appreciate a careful review of the names chosen with an eye towards:

- hierarchy
- singular/plural
- non-CRUD operation naming

Imagine that these names will translate verbatim into the CLI.

The `nexus_tags.txt` file may be the most concise way to consider this.

### CODE REVIEW

File: nexus/src/external_api/http_entrypoints.rs [hunk 1]
```diff
@@ -597,13 +596,15 @@ async fn silo_saml_idp_fetch(
     apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
 
+// TODO: no DELETE for identity providers?
```
Comment (ahl) [inline nexus/src/external_api/http_entrypoints.rs:599]: @davepacheco thoughts?

File: nexus/src/external_api/http_entrypoints.rs [hunk 2]
```diff
@@ -1255,13 +1257,15 @@ async fn ip_pool_ranges_add(
     apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
 
+// TODO should this be a DELETE on a path that includes the range_id?
+// TODO are we missing a GET on a particular range_id?
```
Comment (ahl) [inline nexus/src/external_api/http_entrypoints.rs:1261]: @bnaecker what do you think?

File: nexus/src/external_api/http_entrypoints.rs [hunk 3]
```diff
@@ -3533,6 +3537,7 @@ async fn builtin_users_get_user(
     apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
 
+// TODO is this the right interface?
```
[the remark this answers is not in the document]
Comment (david-crespo) [reply nexus/src/external_api/http_entrypoints.rs:3540]: a) I don't think so, b) probably not. I believe @iliana is actively working on this

File: nexus/tests/output/nexus_tags.txt [hunk 4]
```diff
@@ -76,115 +78,98 @@ timeseries_schema_get                    /timeseries/schema
...
+API operations found with tag "silos"
 OPERATION ID                             URL PATH
-hardware_sleds_get                       /hardware/sleds
-hardware_sleds_get_sled                  /hardware/sleds/{sled_id}
+silo_create                              /silos
+silo_identity_provider_create            /silos/{silo_name}/saml_identity_providers
+silo_identity_provider_list              /silos/{silo_name}/identity_providers
+silo_identity_provider_view              /silos/{silo_name}/saml_identity_providers/{provider_name}
+silo_list                                /silos
+silo_list_users                          /users
```
Comment (david-crespo) [inline nexus/tests/output/nexus_tags.txt:128]: wonder if we should rename `/users` to `/silo_users` so it's immediately apparent that it's not a global list

File: nexus/src/external_api/http_entrypoints.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iliana) [reply nexus/src/external_api/http_entrypoints.rs:3540]: I don't think this is the right abstraction and I wouldn't want to ship a production rack with this endpoint, personally. (I'm currently writing an RFD on the topic of metrics that are associated with resources we already have endpoints for)

File: nexus/src/external_api/http_entrypoints.rs [hunk 5]
```diff
@@ -1231,13 +1232,14 @@ async fn ip_pool_ranges_get(
     apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
 
+// TODO shouldn't this just be a POST on /ip-pools/{pool_name}/ranges?
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/external_api/http_entrypoints.rs:1235]: Dave, Robert, and I talked a bit about this. It seemed distinct enough from the standard REST API to warrant a POST on specific endpoints, similar to instance reboot or disk attach. Part of it was that the ranges don't have names like the rest of the resources, they're identified by a UUID and, currently by the start/end address. POST-ing to `../ranges` to create a new one seems OK. DELETE would require either that clients use the UUID, or accepting a body, which was why we opted for POST.

Another thing pushing us this way was that potentially, adding or removing a range is not a "simple" add / remove operation. Addition could be expanded to include _merging_ neighboring ranges. Deletion could include splitting ranges. The motivation for the latter was Robert pointing out that sometimes one wants to claw back a specific IP from within a range, and requiring that to be a DELETE followed by two POSTs seemed problematic.

I'm not dogmatic about this sort of thing though, whatever makes the most sense. @davepacheco  Did I capture that well enough.

File: nexus/src/external_api/http_entrypoints.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/src/external_api/http_entrypoints.rs:3540]: @iliana I was unaware of the work you were doing here: glad you're taking this on!

Should I move this into another tag or something to clearly denote that it's a temporary interface? Or should I just ignore it?

[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/external_api/http_entrypoints.rs:3540]: I'd put a TODO about deleting it, but otherwise ignore it. There are some tests related to it, and possibly other things I've forgotten about, that would make just deleting it fail.

File: nexus/tests/output/nexus_tags.txt (hunk 4 above)
Comment (davepacheco) [reply nexus/tests/output/nexus_tags.txt:128]: For the record, there's a bunch of discussion about this in chat right now and I think we probably want to open it for other folks that might have thoughts.  I think there's a broader question of how much end users are expected to know about Silos and also what they're allowed to know.

Comment (zephraph) [reply nexus/tests/output/nexus_tags.txt:128]: Yeah, I both agree and disagree with the rename. In the context of the existing patterns here it's quite obvious that the rename would make sense. Or rather that it would be shaped like `/silos/{silo_name}/users`. Contextually though having a `/users` route that _also_ exists (which an operation ID that matches) would be beneficial from an end user perspective. Hopefully after the operator API discussion today we can hone down on how to shape this.

File: nexus/src/external_api/http_entrypoints.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iliana) [reply nexus/src/external_api/http_entrypoints.rs:3540]: (Maybe also a GH issue so we can remember it there, too.)

[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/external_api/http_entrypoints.rs:3540]: https://github.com/oxidecomputer/omicron/issues/1331

File: nexus/src/external_api/http_entrypoints.rs [hunk 6]
```diff
@@ -3445,7 +3449,7 @@ async fn sagas_get_saga(
     path = "/users",
     tags = ["silos"],
 }]
-async fn silo_users_get(
+async fn silo_list_users(
```
Comment (zephraph) [inline nexus/src/external_api/http_entrypoints.rs:3452]: If we decide to go for `/users` I think this would make more sense as `users_list`. Given SDKs and CLIs will be shaped by what's generated here that would make more sense.

Comment (david-crespo) [reply nexus/src/external_api/http_entrypoints.rs:3452]: Yeah, this is probably better than my suggestion of changing the path. The real issue was just the inconsistency.

File: nexus/tests/output/nexus_tags.txt (hunk 4 above)
Comment (zephraph) [reply nexus/tests/output/nexus_tags.txt:128]: My current understanding of this is that every user exists in a silo. Given that, a lot of top level calls happen in the context of the current silo. `/organizations/`, `/session/me`, etc. Some things do not like `/hardware`, and `/images`. 

I think we're lacking clarity here that could be solved in a couple of ways:
1. (my preference) would be to have a set of top routes that are understood to be outside the context of a silo. `fleet` and `silos` are the most obvious to me. `/fleet/hardware` and  `/fleet/images` are some examples which I think help clarify the scope of certain resources. Otherwise it would be assumed that top level routes are in the context of the silo in which the current user is authenticated for. 
2. We make everything very explicit. e.g. `/silos/{silo}/organizations/{org}/projects/{project}`. In that way we'd assume that everything is defaulted explicitly to the fleet level. While this may be the clearest in that global really does mean global and what you see is what you get from the URL perspective it also makes for an odious experience when you're just trying to do things with a resource on the rack. I know other providers don't have the best of URLs, but we do have an opportunity here to be better than the example they set.

I'm sure there are other solutions that haven't occurred to me. It does seem though that right now we're lacking an overall clarity of at what level a resource actually occupies which seems like an undesirable state.

Comment (zephraph) [reply nexus/tests/output/nexus_tags.txt:128]: Of course, it goes without saying that this is a larger change and not in the scope of this PR, ha.

File: nexus/src/external_api/http_entrypoints.rs (hunk 6 above)
Comment (zephraph) [reply nexus/src/external_api/http_entrypoints.rs:3452]: I don't feel so firmly that I believe this PR should be held up for this. The topic of API organization may warrant its own RFD.

File: nexus/tests/output/nexus_tags.txt (hunk 4 above)
Comment (zephraph) [reply nexus/tests/output/nexus_tags.txt:128]: Notably [RFD 234](https://rfd.shared.oxide.computer/rfd/0234#_silos) says this:

> Identities for users, groups, and service accounts are always scoped to a particular Silo. As a result, there’s no need for Silos to appear in the API URL for Organizations, Projects, etc. This is an important isolation measure: there’s no way for users to even ask about resources in other Silos — they don’t have a way to refer to them.

It seems we're already sort of targeting point 1, we may just need to make it more formalized.

File: nexus/tests/output/nexus_tags.txt [hunk 7]
```diff
@@ -22,48 +24,48 @@ spoof_login                              /login
...
+instance_disk_list                       /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/disks
+instance_list                            /organizations/{organization_name}/projects/{project_name}/instances
+instance_migrate                         /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/migrate
+instance_nic_create                      /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/network-interfaces
+instance_nic_delete                      /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/network-interfaces/{interface_name}
+instance_nic_list                        /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/network-interfaces
+instance_nic_update                      /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/network-interfaces/{interface_name}
+instance_nic_view                        /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/network-interfaces/{interface_name}
```
Comment (karencfv) [inline nexus/tests/output/nexus_tags.txt:53]: Even if it's very long, I think I might go for `instance_network_interfaces_*` instead of `instance_nic_*`. I like the idea of clarity, and the CLI has bash completion so it shouldn't be too much of a pain to type? What do you think?

File: nexus/tests/output/nexus_tags.txt [hunk 8]
```diff
@@ -22,48 +24,48 @@ spoof_login                              /login
...
-project_images_delete_image              /organizations/{organization_name}/projects/{project_name}/images/{image_name}
-project_images_get                       /organizations/{organization_name}/projects/{project_name}/images
-project_images_get_image                 /organizations/{organization_name}/projects/{project_name}/images/{image_name}
-project_images_post                      /organizations/{organization_name}/projects/{project_name}/images
+image_create                             /organizations/{organization_name}/projects/{project_name}/images
+image_delete                             /organizations/{organization_name}/projects/{project_name}/images/{image_name}
+image_list                               /organizations/{organization_name}/projects/{project_name}/images
+image_view                               /organizations/{organization_name}/projects/{project_name}/images/{image_name}
 
 API operations found with tag "images:global"
```
Comment (karencfv) [inline nexus/tests/output/nexus_tags.txt:32]: Maybe unrelated to this PR, but why is the tag "images:global" and not "global_images"?

File: nexus/tests/output/nexus_tags.txt [hunk 9]
```diff
@@ -76,115 +78,98 @@ timeseries_schema_get                    /timeseries/schema
...
+vpc_router_create                        /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers
+vpc_router_create_route                  /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}/routes
+vpc_router_delete                        /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}
+vpc_router_delete_route                  /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}/routes/{route_name}
+vpc_router_list                          /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers
+vpc_router_list_routes                   /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}/routes
+vpc_router_update                        /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}
+vpc_router_update_route                  /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}/routes/{route_name}
```
Comment (karencfv) [inline nexus/tests/output/nexus_tags.txt:164]: Same as above

```suggestion
vpc_router_route_update                  /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}/routes/{route_name}
```

**Review by karencfv — COMMENTED** [review]
This is looking really great! It makes me so happy to see these changes, thanks for pushing for semantic naming! 🎉

File: nexus/tests/output/nexus_tags.txt [hunk 10]
```diff
@@ -1,14 +1,16 @@
 API operations found with tag "disks"
 OPERATION ID                             URL PATH
-project_disks_delete_disk                /organizations/{organization_name}/projects/{project_name}/disks/{disk_name}
-project_disks_get                        /organizations/{organization_name}/projects/{project_name}/disks
-project_disks_get_disk                   /organizations/{organization_name}/projects/{project_name}/disks/{disk_name}
-project_disks_post                       /organizations/{organization_name}/projects/{project_name}/disks
+disk_create                              /organizations/{organization_name}/projects/{project_name}/disks
+disk_delete                              /organizations/{organization_name}/projects/{project_name}/disks/{disk_name}
+disk_list                                /organizations/{organization_name}/projects/{project_name}/disks
+disk_view                                /organizations/{organization_name}/projects/{project_name}/disks/{disk_name}
```
[the remark this answers is not in the document]
Comment (zephraph) [reply nexus/tests/output/nexus_tags.txt:6]: Not that it necessarily matters one way or the other, but GitHub's CLI uses `view` and `edit`. Regardless, it's all an upgrade to what we have today, ha.

[the remark this answers is not in the document]
Comment (david-crespo) [reply nexus/tests/output/nexus_tags.txt:6]: @karencfv we had agreed on `edit` -> `update` but were 50-50 on `view` vs. `get`. I guess I slightly prefer `get` since we might as well match the http method when it's natural to do so.

File: nexus/tests/output/nexus_tags.txt [hunk 11]
```diff
@@ -22,48 +24,48 @@ spoof_login                              /login
...
-project_instances_get                    /organizations/{organization_name}/projects/{project_name}/instances
-project_instances_get_instance           /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}
-project_instances_instance_reboot        /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/reboot
-project_instances_instance_serial_get    /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/serial
-project_instances_instance_start         /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/start
-project_instances_instance_stop          /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/stop
-project_instances_migrate_instance       /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/migrate
-project_instances_post                   /organizations/{organization_name}/projects/{project_name}/instances
+instance_console                         /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/serial
```
[the remark this answers is not in the document]
Comment (david-crespo) [reply nexus/tests/output/nexus_tags.txt:41]: yeah I'm with Justin. `/serial-console` route and `instance_serial_console` operation name.

File: nexus/tests/output/nexus_tags.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:6]: I actually took the deliberate step **not** to match the HTTP method because I thought it might be confusing if some matched while others did not. That said: not a strong preference so if someone wants to make a case for one vs. the other that would be clarifying.

File: nexus/tests/output/nexus_tags.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:41]: I'll make those changes.

File: nexus/tests/output/nexus_tags.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (david-crespo) [reply nexus/tests/output/nexus_tags.txt:6]: I can definitely see that side too. It makes you assume things about the semantics that we may not want to imply. Or maybe we actually do because we want these to be spec compliant gets.

File: nexus/tests/output/nexus_tags.txt [hunk 12]
```diff
@@ -22,48 +24,48 @@ spoof_login                              /login
...
+instance_create                          /organizations/{organization_name}/projects/{project_name}/instances
+instance_delete                          /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}
+instance_disk_attach                     /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/disks/attach
+instance_disk_detach                     /organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/disks/detach
```
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:45]: in general we seem to use singular names for things like "instances". I'm neutral here, but chose the singular for what I perceived as consistency.

File: nexus/tests/output/nexus_tags.txt (hunk 7 above)
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:53]: ok cool; I'll make that change

File: nexus/tests/output/nexus_tags.txt (hunk 8 above)
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:32]: yeah; let's set that aside? I agree though that it's ... odd

File: nexus/tests/output/nexus_tags.txt [hunk 13]
```diff
@@ -22,48 +24,48 @@ spoof_login                              /login
...
-ip_pools_delete_ip_pool                  /ip-pools/{pool_name}
-ip_pools_get                             /ip-pools
-ip_pools_get_ip_pool                     /ip-pools/{pool_name}
-ip_pools_post                            /ip-pools
-ip_pools_put_ip_pool                     /ip-pools/{pool_name}
+ip_pool_create                           /ip-pools
+ip_pool_delete                           /ip-pools/{pool_name}
+ip_pool_list                             /ip-pools
+ip_pool_range_create                     /ip-pools/{pool_name}/ranges/add
+ip_pool_range_delete                     /ip-pools/{pool_name}/ranges/delete
```
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:65]: see https://github.com/oxidecomputer/omicron/pull/1329#pullrequestreview-1025399414

File: nexus/tests/output/nexus_tags.txt [hunk 14]
```diff
@@ -76,115 +78,98 @@ timeseries_schema_get                    /timeseries/schema
...
-
-API operations found with tag "racks"
-OPERATION ID                             URL PATH
-hardware_racks_get                       /hardware/racks
-hardware_racks_get_rack                  /hardware/racks/{rack_id}
+project_create                           /organizations/{organization_name}/projects
+project_delete                           /organizations/{organization_name}/projects/{project_name}
+project_list                             /organizations/{organization_name}/projects
+project_update                           /organizations/{organization_name}/projects/{project_name}
+project_update_policy                    /organizations/{organization_name}/projects/{project_name}/policy
```
[the remark this answers is not in the document]
Comment (david-crespo) [reply nexus/tests/output/nexus_tags.txt:100]: I'm with Karen on the operation IDs as operation IDs, but your comment suggests this choice is (for you) more about determining what they will translate to in the CLI. I don't think `oxide update-policy` is clearly better than `oxide policy update` (I would lean toward consistency with whatever else we have), but I also don't have as clear a picture in mind as either of you of what the rules will be for translating from operation ID to CLI command.

File: nexus/tests/output/nexus_tags.txt [hunk 15]
```diff
@@ -76,115 +78,98 @@ timeseries_schema_get                    /timeseries/schema
...
+snapshot_delete                          /organizations/{organization_name}/projects/{project_name}/snapshots/{snapshot_name}
+snapshot_list                            /organizations/{organization_name}/projects/{project_name}/snapshots
+snapshot_view                            /organizations/{organization_name}/projects/{project_name}/snapshots/{snapshot_name}
 
 API operations found with tag "system"
 OPERATION ID                             URL PATH
-builtin_users_get                        /users_builtin
-builtin_users_get_user                   /users_builtin/{user_name}
+system_user_list                         /users_builtin
+system_user_view                         /users_builtin/{user_name}
```
[the remark this answers is not in the document]
Comment (david-crespo) [reply nexus/tests/output/nexus_tags.txt:144]: Yes

File: nexus/tests/output/nexus_tags.txt [hunk 16]
```diff
@@ -76,115 +78,98 @@ timeseries_schema_get                    /timeseries/schema
...
-project_vpcs_post                        /organizations/{organization_name}/projects/{project_name}/vpcs
-project_vpcs_put_vpc                     /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}
+vpc_create                               /organizations/{organization_name}/projects/{project_name}/vpcs
+vpc_delete                               /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}
+vpc_firewall_list_rules                  /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/firewall/rules
+vpc_firewall_update_rules                /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/firewall/rules
+vpc_list                                 /organizations/{organization_name}/projects/{project_name}/vpcs
+vpc_router_create                        /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers
+vpc_router_create_route                  /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}/routes
```
[the remark this answers is not in the document]
Comment (david-crespo) [reply nexus/tests/output/nexus_tags.txt:158]: Consistency!

File: nexus/tests/output/nexus_tags.txt [hunk 17]
```diff
@@ -76,115 +78,98 @@ timeseries_schema_get                    /timeseries/schema
...
-OPERATION ID                             URL PATH
-hardware_racks_get                       /hardware/racks
-hardware_racks_get_rack                  /hardware/racks/{rack_id}
+project_create                           /organizations/{organization_name}/projects
+project_delete                           /organizations/{organization_name}/projects/{project_name}
+project_list                             /organizations/{organization_name}/projects
+project_update                           /organizations/{organization_name}/projects/{project_name}
+project_update_policy                    /organizations/{organization_name}/projects/{project_name}/policy
+project_view                             /organizations/{organization_name}/projects/{project_name}
+project_view_policy                      /organizations/{organization_name}/projects/{project_name}/policy
```
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:102]: as above

File: nexus/tests/output/nexus_tags.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (karencfv) [reply nexus/tests/output/nexus_tags.txt:6]: I'm on the fence about this one as well 😅, but I think I'm leaning more towards the use of `get`. It's a tad more consistent with what people will be used to seeing in SDKs generally. That said, while `get` appears everywhere, there are also weird inconsistencies everywhere. Like for example, AWS uses some weird combination of ["sometimes `get`, sometimes `describe`"](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Operations.html). Both [Azure](https://docs.microsoft.com/en-us/rest/api/compute/disks) and [GCP](https://cloud.google.com/compute/docs/reference/rest/v1) have the more consistent APIs/SDKs and [VMware](https://developer.vmware.com/apis/vmc/latest/operation-index/) is all over the place. So , I guess sticking to `view` isn't a huge deal as long as we don't deviate.

TLDR: I have slight preference for `get` but not a huge deal if we stick with `view` :)

File: nexus/tests/output/nexus_tags.txt (hunk 8 above)
Comment (karencfv) [reply nexus/tests/output/nexus_tags.txt:32]: sounds good!

File: nexus/tests/output/nexus_tags.txt (hunk 12 above)
[the remark this answers is not in the document]
Comment (karencfv) [reply nexus/tests/output/nexus_tags.txt:44]: Yeah, makes sense. Let's keep it singular then

File: nexus/tests/output/nexus_tags.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (karencfv) [reply nexus/tests/output/nexus_tags.txt:100]: I generally like to keep actions (create, delete, attach, import... etc) as subcommands to the resource. This imo keeps the amount of commands at a reasonable amount when listing them as more endpoints are introduced, helps when navigating the `--help` output as a new user, as well as making autocomplete easier :)

This means I'll generally go for
```console
$ oxide project --help
USAGE:
    oxide project [OPTIONS] <SUBCOMMAND>

OPTIONS:
<...>

SUBCOMMANDS:
    create    Create a new project.
    delete    Delete project.
    edit      Edit project settings.
    help      Print this message or the help of the given subcommand(s)
    list      List projects.
    policy    Manage policies.
    view      View project.
<...>

$ oxide project policy --help
USAGE:
    oxide project policy [OPTIONS] <SUBCOMMAND>

OPTIONS:
<...>

SUBCOMMANDS:
    view      View a policy.
    update    Update a policy.
<...>
```

instead of
```console
$ oxide project --help
USAGE:
    oxide project [OPTIONS] <SUBCOMMAND>

OPTIONS:
<...>

SUBCOMMANDS:
    create        Create a new project.
    delete        Delete project.
    help          Print this message or the help of the given subcommand(s)
    list          List projects.
    update        Update project settings.
    update-policy Update a project policy
    view          View project.
    view-policy   View a project policy
<Some other commands for new endpoints that don't exist yet...>
```

File: nexus/tests/output/nexus_tags.txt (hunk 4 above)
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:128]: From this discussion, I'm going to set the `operationId` for `/users` to simply `user_list` (it will imply the current silo for the active user).

To add to the list of outside-of-silos things: silos themselves, hardware, global images, **also** some external networking config (not sure where we configure this), and updates.

File: nexus/tests/output/nexus_tags.txt (hunk 8 above)
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:32]: Maybe we should have a single tag for fleet-wide operations?

File: nexus/src/external_api/http_entrypoints.rs (hunk 2 above)
Comment (ahl) [reply nexus/src/external_api/http_entrypoints.rs:1260]: See above

File: nexus/src/external_api/http_entrypoints.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/src/external_api/http_entrypoints.rs:1234]: That makes sense. Thanks. How about this:

```diff
 ip_pool_create                           /ip-pools
 ip_pool_delete                           /ip-pools/{pool_name}
 ip_pool_list                             /ip-pools
-ip_pool_range_create                     /ip-pools/{pool_name}/ranges/add
-ip_pool_range_delete                     /ip-pools/{pool_name}/ranges/delete
+ip_pool_range_add                        /ip-pools/{pool_name}/ranges/add
 ip_pool_range_list                       /ip-pools/{pool_name}/ranges
+ip_pool_range_remove                     /ip-pools/{pool_name}/ranges/remove
 ip_pool_update                           /ip-pools/{pool_name}
 ip_pool_view                             /ip-pools/{pool_name}
```

In short: I changed `delete` to `remove` to avoid the confusion

File: nexus/tests/output/nexus_tags.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:100]: Making this change

File: nexus/tests/output/nexus_tags.txt [hunk 18]
```diff
@@ -76,115 +78,98 @@ timeseries_schema_get                    /timeseries/schema
...
+vpc_router_update                        /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}
+vpc_router_update_route                  /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}/routes/{route_name}
+vpc_router_view                          /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}
+vpc_router_view_route                    /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/routers/{router_name}/routes/{route_name}
+vpc_subnet_create                        /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/subnets
+vpc_subnet_delete                        /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/subnets/{subnet_name}
+vpc_subnet_list                          /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/subnets
+vpc_subnet_list_nics                     /organizations/{organization_name}/projects/{project_name}/vpcs/{vpc_name}/subnets/{subnet_name}/network-interfaces
```
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt]: This one I think is worth the inconsistency: the network interfaces aren't an object underneath the subnet, and we don't have the full range of CRUD operations here. It's merely a way to view connections from this subnet to network interfaces and instances.

[the remark this answers is not in the document]
Comment (karencfv) [reply nexus/tests/output/nexus_tags.txt]: Ok! that makes sense

**Comment by karencfv** [comment]
Heya @ahl, I was just trying out this fancy new nexus.json file with oxide.rs and the CLI and it turns out the CLI broke in many more unexpected ways than just the method names 😅. Since the testing isn't complete in both CLI and SDKs. What do you think about merging this PR after the demo? I'm totally OK with breaking all the things after that date. Would it be too annoying to do that?

@zephraph @david-crespo what do you think?

**Comment by zephraph** [comment]
I'm on the fence. I think it goes without saying that if this puts the demo at risk then we'd have no choice but to delay it. That said, having the SDKs simplified for when Disney really takes a peek would be ideal. If it's a question of effort I'd prefer have a few people take a little time to help through the migration. That's only my initial thoughts though. I'll spend some time tonight testing it myself and try to come back with a stronger recommendation tomorrow.

**Comment by david-crespo** [comment]
This is a big customer-facing improvement across the board besides that, so it’s probably worth spending a couple of hours to see if maybe it’s less of a disaster in the client libs than it seems at first.

**Comment by zephraph** [comment]
I wish we could just push through with this too, but... I'm not sure we can. Without a larger team effort to tackle this migration @karencfv is right. 

There's a lot that would break. 
1. The current rust SDK (we could switch, but there's progenitor work that still has to be finished so that's effectively blocked)
2. The [proc macro](https://github.com/oxidecomputer/cli/blob/main/cli-macro-impl/src/lib.rs) that generates a lot of the CLI
3. The Go SDK

We've got a lot of conditional logic around naming that makes the SDKs pretty fragile to changes like this.

**Comment by karencfv** [comment]
I already spent a bit more than a couple of hours on this unfortunately 😅  Like Justin says a lot would break 😢 

There are two major pain points regarding the Rust SDK/CLI 
- This proc macro https://github.com/oxidecomputer/cli/blob/main/cli-macro-impl/src/lib.rs . It's highly customised to the previous tags with many conditionals based on the tags unfortunately. 
- This types template https://github.com/oxidecomputer/oxide.rs/blob/main/generator/src/types_templates.rs which and Also, the testing coverage is not optimal, so testing does not guarantee that commands will work if we managed to update the CLI on time.

Also, The work with progenitor is quite far from working with the CLI yet, and will not be able to work with the current CLI generator at all.

**Comment by karencfv** [comment]
There's also the point to consider that given the CLI will change, is it a good use of time to spend that much time updating it?

**Comment by david-crespo** [comment]
Ok, that is persuasive. Seems like we should do the Progenitor stuff first so this change is less painful. On the Go SDK side since there’s no progenitor maybe we could at least reduce the amount of custom renaming logic first.

**Comment by karencfv** [comment]
Yep! Adam and I are working on the progenitor stuff now (from both sides of the fence 😄 ). I'm giving this top priority. The Go SDK will wait a bit as the Terraform provider is in demo condition anyway as there are no update endpoints for disks and instances

File: nexus/tests/output/nexus_tags.txt (hunk 8 above)
Comment (karencfv) [reply nexus/tests/output/nexus_tags.txt:32]: That sounds really good!

**Comment by ahl** [comment]
The latest on the demo is that it may not be for another month. My hope for the demo was to have the API trending towards the state we want it to be in. I'm not clear on when the API/CLI/SDKs may be in the hands of early users, but I expect it will follow close on the heels of the demo. After remote access will come our own robust internal use, documentation, automation, etc. And then product launch.

We will continue to expand and evolve the API. My hope was that by changing its trajectory now there would be less to tidy up later. That is to say: we follow the examples we see so if we can dominate the API with examples we want rather than examples we don't we will follow the positive examples.

In short, I think we need to find a path forward. Here's a simple kludge: I can extract from this PR a one-to-one list of `operationId` name changes. `oxide.rs` and others can pump nexus.json through sed to get the old names. Omicron can move forward and we can make progress with progenitor and the CLI.

To be clear, there are going to be aspects of this that suck and it's going to be brittle. But I think the alternative may be to incur more change and more risk later in the product development lifecycle. I'd suggest interested parties meet tomorrow afternoon PT/morning NZ to discuss.

**Comment by karencfv** [comment]
Happy to catch up in my morning, but just want to comment that I'm perfectly fine with:

> In short, I think we need to find a path forward. Here's a simple kludge: I can extract from this PR a one-to-one list of operationId name changes. oxide.rs and others can pump nexus.json through sed to get the old names. Omicron can move forward and we can make progress with progenitor and the CLI.

I'd really hate to stop this work from moving along, and this seems like a reasonable temporary solution while the new SDK/CLI is up and running :)

**Comment by zephraph** [comment]
@ahl, yep, that sounds like a solid path forward.

File: nexus/src/external_api/http_entrypoints.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/external_api/http_entrypoints.rs]: Yeah, that sounds like a good approach, thanks!

File: nexus/tests/output/nexus_tags.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (ahl) [reply nexus/tests/output/nexus_tags.txt:6]: I'm going to go with `_view`. I think of all the things to change in the future, `s/_view/_get/` is going to be pretty easy in API/SDK/CLI/etc.

**Review by karencfv — APPROVED** [review]
Thanks @ahl !! That list will keep the CLI alive until we have the new SDK
